// @sequelize/core/lib/dialects/mariadb/data-types.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var data_types_exports = {};
__export(data_types_exports, {
  BIGINT: () => import_data_types2.BIGINT,
  BOOLEAN: () => import_data_types2.BOOLEAN,
  DATE: () => import_data_types2.DATE,
  DECIMAL: () => import_data_types2.DECIMAL,
  DOUBLE: () => import_data_types2.DOUBLE,
  ENUM: () => import_data_types2.ENUM,
  FLOAT: () => import_data_types2.FLOAT,
  GEOMETRY: () => GEOMETRY,
  INTEGER: () => import_data_types2.INTEGER,
  JSON: () => import_data_types3.JSON,
  MEDIUMINT: () => import_data_types2.MEDIUMINT,
  REAL: () => import_data_types2.REAL,
  SMALLINT: () => import_data_types2.SMALLINT,
  TINYINT: () => import_data_types2.TINYINT,
  UUID: () => import_data_types2.UUID
});
module.exports = __toCommonJS(data_types_exports);
var import_data_types = require("../mysql/data-types.js");
var import_data_types2 = require("../mysql/data-types.js");
var import_data_types3 = require("../abstract/data-types.js");
class GEOMETRY extends import_data_types.GEOMETRY {
  toSql() {
    var _a;
    const sql = ((_a = this.options.type) == null ? void 0 : _a.toUpperCase()) || "GEOMETRY";
    if (this.options.srid) {
      return `${sql} REF_SYSTEM_ID=${this.options.srid}`;
    }
    return sql;
  }
}
