// bnf-parser/bin/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReferenceRange = exports.Reference = exports.ParseError = exports.SyntaxNode = exports.Compile = exports.Parser = exports.BNF = void 0;
const path = require("path");
const fs = require("fs");
const syntax_1 = require("./syntax");
Object.defineProperty(exports, "SyntaxNode", { enumerable: true, get: function () { return syntax_1.SyntaxNode; } });
Object.defineProperty(exports, "ParseError", { enumerable: true, get: function () { return syntax_1.ParseError; } });
Object.defineProperty(exports, "Reference", { enumerable: true, get: function () { return syntax_1.Reference; } });
Object.defineProperty(exports, "ReferenceRange", { enumerable: true, get: function () { return syntax_1.ReferenceRange; } });
const compiler_1 = require("./compiler");
Object.defineProperty(exports, "Compile", { enumerable: true, get: function () { return compiler_1.Compile; } });
const parser_1 = require("./parser");
Object.defineProperty(exports, "Parser", { enumerable: true, get: function () { return parser_1.Parser; } });
const BNF = new parser_1.Parser(JSON.parse(fs.readFileSync(path.join(__dirname, '../bnf.json'), 'utf8')));
exports.BNF = BNF;
