// @sequelize/core/lib/dialects/sqlite/query-generator-typescript.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var query_generator_typescript_exports = {};
__export(query_generator_typescript_exports, {
  SqliteQueryGeneratorTypeScript: () => SqliteQueryGeneratorTypeScript
});
module.exports = __toCommonJS(query_generator_typescript_exports);
var import_check = require("../../utils/check");
var import_join_sql_fragments = require("../../utils/join-sql-fragments");
var import_string = require("../../utils/string");
var import_query_generator_typescript = require("../abstract/query-generator-typescript");
var import_query_generator = require("../mysql/query-generator");
const REMOVE_INDEX_QUERY_SUPPORTED_OPTIONS = /* @__PURE__ */ new Set(["ifExists"]);
class SqliteQueryGeneratorTypeScript extends import_query_generator.MySqlQueryGenerator {
  describeTableQuery(tableName) {
    return `PRAGMA TABLE_INFO(${this.quoteTable(tableName)})`;
  }
  showIndexesQuery(tableName) {
    return `PRAGMA INDEX_LIST(${this.quoteTable(tableName)})`;
  }
  getToggleForeignKeyChecksQuery(enable) {
    return `PRAGMA foreign_keys = ${enable ? "ON" : "OFF"}`;
  }
  removeIndexQuery(tableName, indexNameOrAttributes, options) {
    if (options) {
      (0, import_check.rejectInvalidOptions)(
        "removeIndexQuery",
        this.dialect.name,
        import_query_generator_typescript.REMOVE_INDEX_QUERY_SUPPORTABLE_OPTIONS,
        REMOVE_INDEX_QUERY_SUPPORTED_OPTIONS,
        options
      );
    }
    let indexName;
    if (Array.isArray(indexNameOrAttributes)) {
      const table = this.extractTableDetails(tableName);
      indexName = (0, import_string.generateIndexName)(table, { fields: indexNameOrAttributes });
    } else {
      indexName = indexNameOrAttributes;
    }
    return (0, import_join_sql_fragments.joinSQLFragments)([
      "DROP INDEX",
      (options == null ? void 0 : options.ifExists) ? "IF EXISTS" : "",
      this.quoteIdentifier(indexName)
    ]);
  }
  getForeignKeyQuery(tableName, columnName) {
    if (columnName) {
      throw new Error(`Providing a columnName in getForeignKeyQuery is not supported by ${this.dialect.name}.`);
    }
    const escapedTable = this.escapeTable(tableName);
    return (0, import_join_sql_fragments.joinSQLFragments)([
      "SELECT id as `constraintName`,",
      `${escapedTable} as \`tableName\`,`,
      "pragma.`from` AS `columnName`,",
      "pragma.`table` AS `referencedTableName`,",
      "pragma.`to` AS `referencedColumnName`,",
      "pragma.`on_update`,",
      "pragma.`on_delete`",
      `FROM pragma_foreign_key_list(${escapedTable}) AS pragma;`
    ]);
  }
  escapeTable(tableName) {
    const table = this.extractTableDetails(tableName);
    if (table.schema) {
      return this.escape(`${table.schema}${table.delimiter}${table.tableName}`);
    }
    return this.escape(table.tableName);
  }
}
